### PR TITLE
Add GPU-deterministic tf.nn.bias_add

### PR DESCRIPTION
**Pull request by duncanriach**
After addressing cuDNN non-determinism in TensorFlow (PRs [24747](https://github.com/tensorflow/tensorflow/pull/24747), [25269](https://github.com/tensorflow/tensorflow/pull/25269), [25796](https://github.com/tensorflow/tensorflow/pull/25796), [29667](https://github.com/tensorflow/tensorflow/pull/29667), [31389](https://github.com/tensorflow/tensorflow/pull/31389)), `tf.nn.bias_add` is the next-most-common source of GPU non-determinism for deep learning models. After this PR, I intend to address other sources of non-determinism before returning to implement a deterministic `bias_add` solution at the CUDA kernel level for higher performance TF determinism when running on GPUs. What this PR does:

* Add (temporary, but not bad) deterministic `tf.nn.bias_add` solution.
* Add `TF_DETERMINISTIC_OPS` environment variable that enables that solution and also enables cuDNN determinism.
* Tests for the above.

### CODE REVIEW

**Comment by duncanriach** [comment]
@rthadur please could you remove the WIP label? This PR is ready for review.

**Comment by aaroey** [comment]
@chsigg it seems you had reviewed changes to python/kernel_tests/cudnn_determinism_test.py in the past, could you help to take a look at this as well?

**Review by chsigg — DISMISSED** [review]
Looks good, thanks a lot!

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by rthadur** [comment]
@duncanriach can you please check build failures ?

**Comment by duncanriach** [comment]
@rthadur The build failures seemed to be unrelated to my changes. I rebased to top-of-tree and force-pushed.

**Review by chsigg — DISMISSED** [review]
Reapprove.

**Comment by duncanriach** [comment]
Hi @rthadur, please will you make the checks run? `Ubuntu CC` and `Ubuntu Sanity` have be in the `Expected` state for seven days.

**Review by chsigg — DISMISSED** [review]

**Comment by duncanriach** [comment]
Hi @rthadur, please will you add the `kokoro:force-run` label. I would like to see if CI passes before requesting another review.

**Comment by rthadur** [comment]
done thank you

**Comment by duncanriach** [comment]
Okay, thank you, @rthadur. `Linux GPU` is now passing. I think that the issue that kept re-occurring before is resolved. However, there are these three build failures in `Android Demo App`, `Windows Bazel`, and `Windows Bazel GPU` that seem to be unrelated to my changes. What is the recommended solution for that?

I think it's now possible to ask @chsigg to re-review. There are virtually no changes from the previous reviews he did. The only thing I did to make the problematic test pass was to add the `@test_util.run_deprecated_v1` decorator to `testDeterministicGradients` in `bias_op_deterministic_test.py`, which is the same as for the other, existing bias_op tests. This prevents the test from running on the TF v2.0 API, in which `gradients_impl.gradients()` is not available and/or in eager mode (via v1.x or v2.x APIs) in which `array_ops.placeholder` does not work.

**Comment by duncanriach** [comment]
FYI, I'm currently attempting to find a recent commit hash to rebase my branch from that will both build locally and also pass all the CI tests. Trying `c33b99e859e70f5acb8d62909ab364a26e60954e`.

**Comment by duncanriach** [comment]
Hi @rthadur, please will you add the `kokoro:force-run` label. Hopefully, I've just rebased to a commit that will pass all the checks.

**Comment by rthadur** [comment]
> Hi @rthadur, please will you add the `kokoro:force-run` label. Hopefully, I've just rebased to a commit that will pass all the checks.

done

**Comment by duncanriach** [comment]
`Android Demo App` is failing with `gcc: error: unrecognized command line option '-std=c++14'`. It seems like this has been recurring for a while and I don't see how it can be related to my changes. Does anyone know if it's possible to just ignore this error?

**Comment by rthadur** [comment]
@chsigg  gentle ping for your review comments ?

**Comment by duncanriach** [comment]
`Windows Bazel` failed because `MultiWorkerMirroredStrategyTest` failed, which failed because `tensorflow.python.framework.errors_impl.UnknownError: Could not start gRPC server`. This does not seem to have any relationship with my changes. Does anyone know if it's possible to just ignore this error?

**Review by chsigg — DISMISSED** [review]

**Review by aaroey — DISMISSED** [review]

File: tensorflow/python/kernel_tests/bias_op_deterministic_test.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    # gradient result, not passing the upstram gradients through the op's
+    # gradient generation graph. This is the reason for using the
+    # gradient_injector_op
+    grad_ys = None
+    bias_gradients_op = gradients_impl.gradients(
+        gradient_injector_op, bias_op, grad_ys=grad_ys,
+        colocate_gradients_with_ops=True)
+    for i in range(5):
+      feed_dict = {upstream_gradients: self._random_ndarray(out_shape)}
+      self._assert_reproducible(bias_gradients_op, feed_dict=feed_dict)
```
Comment (aaroey) [inline tensorflow/python/kernel_tests/bias_op_deterministic_test.py]: This test is failing during the merging, it seems the result is not deterministic, we're still investigating. Do you have any idea?

Comment (duncanriach) [reply tensorflow/python/kernel_tests/bias_op_deterministic_test.py]: I'm wondering if the test is failing when run using XLA and that the XLA kernel fusion is replacing part of all of the deterministic solution with a non-deterministic kernel.

Comment (duncanriach) [reply tensorflow/python/kernel_tests/bias_op_deterministic_test.py]: In `tensorflow/tensorflow.bzl` in the master branch, it looks like `xla_enable_strict_auto_jit` is ignored. I'm wondering if it's not being ignored in the CI setup that the test is failing in. I'm wondering if `xla_enable_strict_auto_jit = True` should be removed from `tensorflow/python/kernel_tests/BUILD` for `bias_op_deterministic_test`. As an experiment, do you think it's worth me pushing a commit to this PR branch to remove it, @aaroey ?

Comment (duncanriach) [reply tensorflow/python/kernel_tests/bias_op_deterministic_test.py]: I  have reproduced what appears to be the same issue in a slightly different context. Enabling auto-JIT does make that test fail. XLA is probably replacing the deterministic mini-graph of ops with the original non-deterministic op kernel. I'll push a commit with auto-JIT disabled in a minute.

Comment (duncanriach) [reply tensorflow/python/kernel_tests/bias_op_deterministic_test.py]: Hi @aaroey, I just disabled XLA JIT for this test in the most recent commit. Please will review and run tests again and see if that resolves the issue.

**Comment by duncanriach** [comment]
Hi @rthadur, please will you force-run the CI tests?

**Comment by rthadur** [comment]
Done

**Comment by duncanriach** [comment]
@aaroey: Gentle nudge. I have marked the failing test not to run with XLA JIT; that should solve the problem. **Please will you try to merge again?**

None of the current CI failures are related to this PR.

**Review by aaroey — DISMISSED** [review]
Thanks for the fix, will try to merge again.

File: tensorflow/python/kernel_tests/BUILD [hunk 2]
```diff
@@ -1665,18 +1683,36 @@ cuda_py_test(
...
         "//tensorflow/python:gradients",
         "//tensorflow/python:nn_grad",
         "//tensorflow/python:nn_ops",
+        "//third_party/py/numpy",
+    ],
+)
+
+cuda_py_test(
+    name = "bias_op_deterministic_test",
+    size = "small",
```
Comment (aaroey) [inline tensorflow/python/kernel_tests/BUILD]: This and the bias_op_test test below are timing out, please change the size to 'medium'.

**Review by aaroey — APPROVED** [review]

**Pull request merged**